### DISCUSSION TITLE
Tips on breaking change migrations

### DISCUSSION THREAD

**Discussion by kentcdodds** [prisma/prisma#7421] [Q&A]
DBs are not my forte and I'm new to prisma. I'm using Postgres.

Let's say that I have a schema model that says one field is an enum of ['ONE', 'TWO', 'THREE']. The production database fills up with that data, then I decide that 'TWO' isn't a valid value anymore so I change the enum to ['ONE', 'THREE']. The migration will fail because some data in the db has the value 'TWO' and that's no longer permitted. Let's say that I want all 'TWO' values to be changed to 'THREE'.

What's the "proper" way to go about migrating the data? Do I just write a little script and execute it against the production database? Is this something people normally do in their build/deploy pipeline?

**Comment by Boeing787** [comment]
Totally. As you said, writing a script against the prod database is the way to go, but I have a couple of additional suggestions (not Prisma specific, but all of these are valid statements in `Migrate`);

- Find all `TWO`s and change them to `THREE`s
- Rename the old enum to something else (perhaps enum_name_deprecated) using an `ALTER TYPE your_enum_name RENAME TO your_enum_name_deprecated;` statement
- Create a new enum with the original name (`your_enum_name`) with just `ONE` and `THREE` as options
- Alter the column using `your_enum_name_deprecated` to use `your_enum_name` with an `ALTER TABLE ... ALTER COLUMN ... TYPE ... USING ...`
- Now that your column is moved over to `your_enum_name` you can safely `DROP` `your_enum_name_deprecated`;
- I would make sure to run this process inside of a transaction so any errors can be safely rolled back without committing. 

That's the general way to safely migrate an enum; in Postgres, you can _add_ values to an enum, but can't subtract, so to eliminate `TWO`, you'll need to create another enum altogether described above.

This brings me to a word of caution (and a matter of *personal* opinion, so take this with a grain of salt): enums are a pain to work with for this reason; it is difficult to remove a stale or deprecated value from them. If you can use a `text` type accompanied by a `CONSTRAINT` on your column to only permit `text` matching `ONE` and `THREE`, that's a much easier route to take.

Would love to hear other approaches to enums in Prisma and PG. This is just one way of doing things.

**Reply by zwily** [reply]
The main reason I use enum types in Postgres is for easy sorting on that column. I’ve done that migration dance many times! If db-sorting of that column isn’t one of your use cases, a text field with a constraint is way easier.

**Reply by tomhoule** [reply]
As a prisma migrate maintainer, I completely agree with @Boeing787's answer. The tool can only go so far to help you when it comes to schema migrations, the only way to migrate without downtime at scale is small non-breaking changes that your roll out progressively. It's basically the "expand and contract" pattern: create the new thing (column, enum, etc.), then do a data migration (it's best if it's separated from your schema migrations — see [this classic blog post](https://thoughtbot.com/blog/data-migrations-in-rails) that applies to Prisma just as well), and when everything is fine, a last migration to clean up the columns/enums/... you do not need anymore. It's more work, but every step is safer, less risky.

**Reply by kentcdodds** [reply]
Any chance Prisma could help automate some of this?

My thought was to create a new database within the postgres install and have two prisma schemas and clients until the migration is complete. Is that the right approach when applying this pattern with prisma?

**Reply by tomhoule** [reply]
We'd love to. It hasn't been a priority so far because we have issues we see as more urgent, and we also don't have a concrete vision of how this could work, but assisting users to get their migration strategy right is definitely something we are interested in. For the coordination between the client API and schema migrations, I think the `@ignore` and `@@ignore` attributes go a long way (see my other reply below).

**Reply by kentcdodds** [reply]
This comment helped me realize that we may not need a new database at all: https://github.com/prisma/prisma/discussions/7421#discussioncomment-820556

**Comment by ludofleury** [comment]
In SQL world we would transactionally:

1.  update the values with update select statement  (move "two" to the new acceptable value)
2. update the schema

And we could make the migration partially reversible as well, according that the "new acceptable value" would be entirely new and not an existing one. 

**Example**
Fully reversible: "two", become "five", which didn't exist prior to the migration.
Not reversible: "two" become "three", which already existed before, so we wouldn't be able to rollback the migration.

Migration is always a complex pain point in continuous deployment, and in a real world cannot be fully "generated" but the execution could be fully automated.

Most of ORM's will generate the migration and let the developer edit them, version them.
The pipeline are run & execute migrations.

**Comment by kentcdodds** [comment]
@delbaoliveira [suggested](https://twitter.com/delba_oliveira/status/1400216575198863363) this: https://www.prisma.io/dataguide/types/relational/expand-and-contract-pattern (thank you Delba!)

That really looks perfect. Unfortunately it appears Prisma doesn't make doing this very easy. For example, you're client would have to be able to read/write from the old schema and the new schema at the same time and unless I'm mistaken, prisma wouldn't be able to do that due to the (magical ✨) TypeScript type defs.

**Reply by ryands17** [reply]
This should be possible. Taking the scenario provided in the above doc, we can:

1. Add `firstName` and `lastName` columns and make the `name` column optional
2. Populate data over to the new columns whilst keeping the old one
3. Change application logic to start writing to the new columns (doesn't affect TS types)
4. Remove the old `name` column

It's a multitude of steps that can be performed but it should work as expected.

**Reply by kentcdodds** [reply]
Oh, that actually makes sense. I hadn't considered that this could all be done without creating an entirely new database! Maybe that's what @Boeing787 was suggesting and I just didn't understand.

**Reply by tomhoule** [reply]
Note that there is also an [`@ignore` attribute](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#ignore) you can use to make this process easier: you can `@ignore` the old fields after the first deployment. That way you know you weren't using them anymore (because `@ignore` will make them disappear from the Client API) at the later point where you are ready to remove them.

**Reply by kentcdodds** [reply]
Very helpful, thank you!

**Reply by hhimanshu** [reply]
What I am little confused (pardon my little knowledge) is how many migrations we need to run and what goes in each migration?
For example

### Change Set 1
#### Db Migration
1. Add a `new` field
#### Application
1. Update code to write to both `old` and the `new` field.
2. Read from the `old` field (No changes here)


> 🚀 Deploy changes in production 🚀 


### Change Set 2
#### DB Migration
1. Copy the data from `old` field to the `new` field
```
UPDATE "Profile" SET biography = bio;
```
Quoting [Step 3](https://www.prisma.io/docs/guides/migrate/developing-with-prisma-migrate/customizing-migrations#example-use-the-expand-and-contract-pattern-to-evolve-the-schema-without-downtime) from the docs

3. Mark `@ignore` on the `old` field.
#### Application
1. Stop writing to `old` field. Remove all references from code that "writes" to the new field
2. Start reading from the `new` field. By this time, the `new` field will be populated with *all* previous data from `old` column so it is safe reading from `new` field.

> 🚀 Deploy changes in production 🚀 

### Change Set 3
#### DB Migration
1. Remove `old` field from the schema

#### Application
1. Remove all references in the code that refer to `old` field.

> 🚀 Deploy changes in production 🚀 


So, as far as I see, it needs `3` production deployments for implementing **Expand and Contract** pattern. Am I missing anything?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
